### DISCUSSION TITLE
Docker-compose: Hardcoded container port? not documented

### DISCUSSION THREAD

**Discussion by tootbrute** [photoprism/photoprism#2941] [General]
I have been using Photoprism for a few years now. Great program! I recommend it to all my friends.

Recently, there must've been a docker change because one of my photoprism instances stopped working. When I was troubleshooting I noticed that running docker-compose up had a line in the logs saying it was waiting for connections on 0.0.0.0:2342.

In docker-compose.yml, I had defined my ports like this:
```
   ports:
      - "2222:2222" # HTTP port (host:container)
```
After pulling my hair for a bit I decided to try this.
```
   ports:
      - "2222:2342" # HTTP port (host:container)
```
Then it worked!

Maybe there could be a commented added in the docker-compose file? It won't affect newbies but for those running multiple instances it could save some headache.

example:
`# HTTP port (host:container). NOTE: container port must be 2342!`

Thank you!

**Comment by jrstrunk** [comment]
Just checked my local config and yup it didn't work when I set the container port to be something other than `2342`. I think it is a good idea to add this to the config. If you make this change and open a PR, maybe the maintainers will approve. Although I might use weaker language, like "should" instead of "must" in the comment because I'm sure you can get it working with a different port somehow ;)

**Reply by lastzero** [reply]
Thank you for your feedback!

You can add the `expose` service property in your docker-compose.yml to use alternative ports if needed:

- https://docs.docker.com/compose/compose-file/compose-file-v3/#expose

However, only one service should run in each container (based on the concept of Docker) and there is no conflict with the host computer, since you can flexibly assign the container port as shown above:

```yaml
   ports:
      - "2222:2342"
```

~~I wonder what advantage it might provide to change the default container port? That is, apart from having to add a lot of notes in our docs to mention that the actual port may be different and will depend on yet another config option...~~

**Reply by lastzero** [reply]
@jrstrunk My apologies! After reading it again, it seems like your suggestion is to extend the existing inline comment for users unfamiliar with Docker so they don't accidentally change the container port? Sure, that would be possible if it helps and can't be misunderstood somehow so that, for example, users think they can't change the port at all... 👍

**Reply by jrstrunk** [reply]
@lastzero first let me say thank you for the great work you've done on this project. But yes I was just agreeing with @tootbrute that a comment guiding users to not change the docker port number unless they know what they are doing would be helpful I think. I might have changed the container port to make it the same as the host port I changed if I hadn't seen this discussion first! 

But I understand that it is not your responsibility to make sure the users know docker, and if you don't want to add more clutter to the docs then that is just fine too.

**Reply by lastzero** [reply]
You're welcome! If it helps a few users, we should improve our docs.

I've just added more detailed notes on this to the config examples available at [dl.photoprism.app/docker/](https://dl.photoprism.app/docker/):

```yaml
    ports:
      ## Web server port mapping in the format "Host:Container". To use a different port, change the host port
      ## on the left-hand side and keep the container port, e.g. "80:2342" (for HTTP) or "443:2342 (for HTTPS):
      - "2342:2342"
```

Related issue:
- #4017
